Improving python interface for output with HFD5 application

msandre opened the pull request:
- Removed list_of_elements and list_of_conditions as json input parameters.

- Xdmf support (tested with paraview and visIt).

- Output python processes for writing model part (normal or partitioned) with a single mesh for all time steps or a mesh for each time step.

 Simplified class diagrams:

c++
![cpp](https://user-images.githubusercontent.com/22290001/36041881-3b8ad93a-0dca-11e8-8db8-a36142235e55.png)

xdmf
![xdmf](https://user-images.githubusercontent.com/22290001/36041889-436cf228-0dca-11e8-8738-5d274a26490d.png)

output python processes
![hdf5_python_output](https://user-images.githubusercontent.com/22290001/36041902-4e48c73a-0dca-11e8-84c2-2a1dbf96687e.png)

Commit:
small.

Commit:
added xdmf utils.

Commit:
cleaned class structure.

Commit:
added fixed mesh hdf5 output.

Commit:
starting point for xdmf post processing.

Commit:
setting time list from dir contents.

Commit:
added fix for 2 node line elements and comments.

Commit:
small.

Commit:
added hdf5 output frequency.

Commit:
moved sorted coordinates process to cpp.

Commit:
merged master.

Commit:
fixed tests after merge.

Commit:
added hdf5 xdmf connectivities writer process.

Commit:
added fix for trilinos ns solver fractionalstep.

Commit:
changed xdmf class structure.

Commit:
Merge branch 'master' into feature-xdmf-for-hdf5

Commit:
small.

Commit:
added WriteInfo.

Commit:
move data set partitioning to a utility.

Commit:
small.

Commit:
small.

Commit:
made file path args const&.

Commit:
small.

Commit:
changed more value string agrs to const&.

Commit:
added compare element and condition type util.

Commit:
Added Read/Write for string attributes to hdf5 file.

Commit:
added mName to ConnectivitiesData.

Commit:
Merge branch 'master' into feature-xdmf-for-hdf5 to use new core util.

Commit:
added File::GetGroupNames

Commit:
added factor elements and condition utility

Commit:
fix typo.

Commit:
automated extraction of element/condition types from container and restructured/simplified code.

Commit:
reorganized ReadModelPart to base.

Commit:
moved points data and connectivities data.

Commit:
removed list of elements/conditions from output process params.

Commit:
added custom_processes folder and organized.

Commit:
fix xdmf connectivity writer to skip partition dset.

Commit:
removed asserts.

Commit:
reordered python hdf5 output processes.

Commit:
cleanup.

Commit:
making single process for extracting xdmf data.

Commit:
cleaning up python scripts.

Commit:
adding non-partitioned output in new style.

Commit:
Merge branch 'master' into feature-xdmf-for-hdf5 to update compare element functions.

Commit:
update to new comparison utilities from master after merge.

Commit:
small.

loumalouomega commented:
Impressive, nice job

loumalouomega commented on .vscode/settings.json:
```diff
@@ -0,0 +1,45 @@
+{
```
Just one comment, be careful with vscode configuration

msandre replied:
Thanks for spotting that :+1:

Commit:
removed vscode settings.

philbucher commented on applications/HDF5Application/custom_io/hdf5_connectivities_data.cpp:
```diff
@@ -0,0 +1,191 @@
+
+    // Fill arrays and perform checks.
+    #pragma omp parallel for
+    for (unsigned i = 0; i < num_conds; ++i)
```
I am quite sure this does not compile in win, I think omp loop have to use `int`
can you confirm this @jcotela ?

jcotela replied:
@philbucher yes, MSVC uses an older version of OpenMP and only supports int loops.

msandre replied:
fixed

philbucher commented on applications/HDF5Application/custom_io/hdf5_connectivities_data.cpp:
```diff
@@ -0,0 +1,191 @@
+    const ElementType& r_elem = KratosComponents<ElementType>::Get(mName);
+    const unsigned num_new_elems = mIds.size();
+    const unsigned geometry_size = r_elem.GetGeometry().size();
+    KRATOS_ERROR_IF(geometry_size != mConnectivities.size2())
```
I would usually use `KRATOS_ERROR_IF_NOT(... == ...)` in this case but it is also fine like this.
If you want to change it, I haven't marked all the occurences

msandre replied:
I don't know why but somehow this form is more readable to me. Must be a preference thing.

philbucher replied:
fine for me, just a preference

philbucher commented on applications/HDF5Application/custom_io/hdf5_connectivities_data.cpp:
```diff
@@ -0,0 +1,191 @@
+#include "custom_io/hdf5_connectivities_data.h"
+
+#include "includes/kratos_components.h"
+#include "utilities/openmp_utils.h"
```
do you need this?

msandre replied:
I checked each include and it appears I use them all (openmp_utils.h for omp.h).

philbucher commented on applications/HDF5Application/custom_io/hdf5_connectivities_data.cpp:
```diff
@@ -0,0 +1,191 @@
+    if (KratosComponents<ElementType>::Has(mName))
+    {
+        const auto& r_elem = KratosComponents<ElementType>::Get(mName);
+        ws_dim = r_elem.WorkingSpaceDimension();
```
not too descriptive, maybe use the full name?

msandre replied:
true, but the function body is very small and the ws_dim only appear in an assignment or setter operation where WorkingSpaceDimension is explicity written so I chose the shorter form for brevity.

philbucher commented on applications/HDF5Application/custom_io/hdf5_connectivities_data.cpp:
```diff
@@ -0,0 +1,191 @@
+    rFile.WriteDataSet(rPath + "/Ids", mIds, rInfo);
+    rFile.WriteDataSet(rPath + "/PropertiesIds", mPropertiesIds, rInfo);
+    rFile.WriteDataSet(rPath + "/Connectivities", mConnectivities, rInfo);
+    int ws_dim, dim, nnodes;
```
naming is not consistent
Here you use `nnodes` but later you use e.g. `num_new_elemenents`
I suggest to use `num_...`

msandre replied:
changed to num_nodes.

philbucher commented on applications/HDF5Application/custom_io/hdf5_connectivities_data.cpp:
```diff
@@ -0,0 +1,191 @@
+    KRATOS_CATCH("");
+}
+
+void ConnectivitiesData::CreateEntities(NodesContainerType& rNodes,
```
I think these args can all be `const`

msandre replied:
fixed

philbucher commented on applications/HDF5Application/custom_io/hdf5_connectivities_data.cpp:
```diff
@@ -0,0 +1,191 @@
+
+    // Fill arrays and perform checks.
+    #pragma omp parallel for
+    for (unsigned i = 0; i < num_elems; ++i)
```
see previous comment

philbucher commented on applications/HDF5Application/custom_io/hdf5_file.cpp:
```diff
@@ -158,146 +158,180 @@ void File::GetAttributeNames(std::string ObjectPath, std::vector<std::string>& r
-    KRATOS_TRY;
-    KRATOS_ERROR << "Calling the base class method. Please override in the derived class." << std::endl;
-    KRATOS_CATCH("");
+    KRATOS_ERROR_IF(Internals::IsPath(rPath) == false) << "Invalid path: " <<rPath << std::endl;
```
`KRATOS_ERROR_IF_NOT(Internals::IsPath(rPath))`

msandre replied:
done

philbucher commented on applications/HDF5Application/custom_io/hdf5_file.cpp:
```diff
@@ -158,146 +158,180 @@ void File::GetAttributeNames(std::string ObjectPath, std::vector<std::string>& r
     KRATOS_TRY;
-    KRATOS_ERROR << "Calling the base class method. Please override in the derived class." << std::endl;
-    KRATOS_CATCH("");
+    boost::timer timer;
```
I am not sure if it is a good idea to introduce more dependencies on boost 
Doesn't the standard have sth for this?
Or you could use the timer from the openmp_utils

msandre replied:
I think it has somethings that can be used to construct a timer mechanism, but this is something we should define in one place (in core). Until we do this, I would leave this as it is. There is a Timer in the core timer.h, but it's not so suitable for timing a single function imo because it uses global variables, which I would prefer to avoid.

RiccardoRossi replied:
@msandre I also would avoid using boost::timer, since i think it does fake outputs if you have openmp active. 

c++11 has std::chrono if you need a similar feature, however i am not sure of how it behaves when openmp is active.

The OpenMP timer could be a good option too.

Having said this i definitely do not consider this a blocker, so i am just commenting

philbucher commented on applications/HDF5Application/custom_io/hdf5_file.h:
```diff
@@ -376,18 +376,18 @@ void File::ReadAttribute(std::string ObjectPath, std::string Name, TScalar& rVal
         << "H5Sget_simple_extent_ndims failed." << std::endl;
     KRATOS_ERROR_IF(H5Sclose(space_id) < 0) << "H5Sclose failed." << std::endl;
-    KRATOS_ERROR_IF(ndims != 0) << "Attribute \"" << Name << "\" is not scalar." << std::endl;
+    KRATOS_ERROR_IF(ndims != 0) << "Attribute \"" << rName << "\" is not scalar." << std::endl;
```
`KRATOS_ERROR_IF_NOT(ndims == 0)`

msandre replied:
I would leave. As we discussed above, it is basically a style/preference thing.

philbucher commented on applications/HDF5Application/custom_processes/hdf5_xdmf_connectivities_writer_process.cpp:
```diff
@@ -0,0 +1,111 @@
+    KRATOS_ERROR_IF(mpFile->HasPath(node_ids_path) == false)
+        << "Path \"" << node_ids_path << "\" was not found." << std::endl;
+
+    Vector<int> node_ids;
```
`Vector` or `std::vector`?

msandre replied:
Vector is used in the file interface.

philbucher commented on applications/HDF5Application/custom_processes/hdf5_xdmf_connectivities_writer_process.cpp:
```diff
@@ -0,0 +1,111 @@
+    mKratosToXdmfIdTable.resize(num_points);
+    const int num_threads = OpenMPUtils::GetNumThreads();
+    OpenMPUtils::PartitionVector partition;
+    OpenMPUtils::DivideInPartitions(num_points, num_threads, partition);
```
I think this way of doing omp is outdated
What do you say @jcotela ?

msandre replied:
:+1:  in any case it is not the best way to parallelize this loop. fixed.

philbucher commented on applications/HDF5Application/custom_processes/hdf5_xdmf_connectivities_writer_process.cpp:
```diff
@@ -0,0 +1,111 @@
+    KRATOS_TRY;
+
+    std::vector<std::string> labels;
+    Matrix<int> connectivities;
```
`Matrix` or `std::vector<std::vector<int>>`?

msandre replied:
Matrix

msandre replied:
actually it appears to be unused so neither :+1:

philbucher commented on applications/HDF5Application/custom_utilities/factor_elements_and_conditions_utility.cpp:
```diff
@@ -0,0 +1,274 @@
+{
+    KRATOS_TRY;
+
+    const int max_size = 1000;
```
constexpr?

msandre replied:
I think either is fine here but will use constexpr.

philbucher commented on applications/HDF5Application/custom_utilities/factor_elements_and_conditions_utility.cpp:
```diff
@@ -0,0 +1,274 @@
+            ss << r_name << '\n';
+        const std::string send_buf = ss.str();
+
+        // Gather components from each process and construct the global set of
```
If I understand the next loop correctly you are doing some sort of [allreduce](https://www.mpich.org/static/docs/v3.1/www3/MPI_Reduce.html)
First you gather and then you scatter again
Couldn't you use the mpi-cmd directly? It might not be allreduce but I think there should be sth for this.

msandre replied:
I think it is closest to a gather, but with gather we need to have all the messages the same size.

philbucher commented on applications/HDF5Application/custom_utilities/factor_elements_and_conditions_utility.h:
```diff
@@ -0,0 +1,149 @@
+{
+///@name Kratos Classes
+///@{
+
```
maybe some doxygen of what this utility does?

msandre replied:
added

philbucher commented on applications/HDF5Application/custom_utilities/hdf5_data_set_partition_utility.cpp:
```diff
@@ -0,0 +1,71 @@
+
+    const std::string partition_path = rPath + DataSetPartitionUtility::PartitionSuffix;
+    const unsigned my_pid = rFile.GetPID();
+    Vector<int> my_partition;
```
`Vector` or `std::vector`?

msandre replied:
Vector

philbucher commented on applications/HDF5Application/python_scripts/partitioned_multiple_mesh_temporal_output_process.py:
```diff
@@ -0,0 +1,42 @@
+import KratosMultiphysics
```
I think this file could derive from `MultipleMeshTemporalOutputProcess`

msandre replied:
Good point, there is a lot of copying going on here, I will do something about this.

philbucher commented on applications/HDF5Application/python_scripts/partitioned_single_mesh_temporal_output_process.py:
```diff
@@ -0,0 +1,41 @@
+import KratosMultiphysics
```
also this file seems to do some things in common with the files above

msandre replied:
Will be fixed.

philbucher commented on applications/HDF5Application/python_scripts/single_mesh_temporal_output_process.py:
```diff
@@ -0,0 +1,41 @@
+import KratosMultiphysics
```
same for this file

philbucher commented on applications/HDF5Application/tests/test_HDF5Application.py:
```diff
@@ -1,7 +1,9 @@
 import KratosMultiphysics
```
do you intentionally not add the xdmf test to be executed here?

Also I think you could have a `test_HDF5Application_mpi` which populates the mpi-test-suites with the mpi tests

msandre replied:
it was a mistake, thanks for spotting it.

philbucher commented on applications/HDF5Application/tests/test_xdmf_output.py:
```diff
@@ -0,0 +1,161 @@
+        params = Parameters("""
+        {
+            "prefix" : "/ModelData",
+            "list_of_elements" : ["Element2D3N", "Element2D4N"],
```
I believe this is not needed any more right?

msandre replied:
correct

philbucher reviewed (changes requested):
Hi @msandre 
Thx first of all for this amazing work!
I compiled and tested your branch and here are several comments:
- It would be very helpful to have a readme (similar to the EigenSolversApplication) on which version of hdf5 one needs and how to compile etc (serial parallel etc)
also some comments abt the usage would be good
- Take a look at the codacy suggestions, I think some of them are very helpful
- I tried a simple (structural) example, but it was not working bcs `CONSTITUTIVE_LAW` is an unsupported variable type. I think the fluid-guys are starting to use this too, so adding this might be handy (have a look at #1455 )
- the only reason for not approving is the unsigned omp loop which I think does not compile in win. My other comments would not be a blocker for approval
- outlook: maybe we could have a postprocessor-factory in the future which would create the hdf5 or the gid output
- running the tests gives a bunch of output:
~~~sh
 |  /           |             
 ' /   __| _` | __|  _ \   __|
 . \  |   (   | |   (   |\__ \ 
_|\_\_|  \__,_|\__|\___/ ____/
           Multi-Physics 5.3.0-bb15e82
Importing    KratosHDF5Application
Initializing KratosHDF5Application... 
Running cpp unit tests ...
TestHDF5ConnectivitiesData_CreateConditions                              OK.

TestHDF5ConnectivitiesData_CreateElements                                OK.

TestHDF5ConnectivitiesData_ReadData                                      OK.

TestHDF5ConnectivitiesData_SetData1                                      OK.

TestHDF5ConnectivitiesData_SetData2                                      OK.

TestHDF5ConnectivitiesData_WriteData                                     OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_AddPath                                                     OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5G.c line 314 in H5Gcreate2(): unable to create group
    major: Symbol table
    minor: Unable to initialize object
  #001: ../../../src/H5Gint.c line 194 in H5G__create_named(): unable to create and link to group
    major: Symbol table
    minor: Unable to initialize object
  #002: ../../../src/H5L.c line 1638 in H5L_link_object(): unable to create new link to object
    major: Links
    minor: Unable to initialize object
  #003: ../../../src/H5L.c line 1882 in H5L_create_real(): can't insert link
    major: Symbol table
    minor: Unable to insert object
  #004: ../../../src/H5Gtraverse.c line 861 in H5G_traverse(): internal path traversal failed
    major: Symbol table
    minor: Object not found
  #005: ../../../src/H5Gtraverse.c line 780 in H5G_traverse_real(): traversal operator failed
    major: Symbol table
    minor: Can't move to next iterator location
  #006: ../../../src/H5L.c line 1674 in H5L_link_cb(): name already exists
    major: Symbol table
    minor: Object already exists
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5G.c line 314 in H5Gcreate2(): unable to create group
    major: Symbol table
    minor: Unable to initialize object
  #001: ../../../src/H5Gint.c line 194 in H5G__create_named(): unable to create and link to group
    major: Symbol table
    minor: Unable to initialize object
  #002: ../../../src/H5L.c line 1638 in H5L_link_object(): unable to create new link to object
    major: Links
    minor: Unable to initialize object
  #003: ../../../src/H5L.c line 1882 in H5L_create_real(): can't insert link
    major: Symbol table
    minor: Unable to insert object
  #004: ../../../src/H5Gtraverse.c line 861 in H5G_traverse(): internal path traversal failed
    major: Symbol table
    minor: Object not found
  #005: ../../../src/H5Gtraverse.c line 641 in H5G_traverse_real(): traversal operator failed
    major: Symbol table
    minor: Callback failed
  #006: ../../../src/H5L.c line 1674 in H5L_link_cb(): name already exists
    major: Symbol table
    minor: Object already exists
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_CreateGroup                                                 OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5O.c line 246 in H5Oopen(): unable to open object
    major: Symbol table
    minor: Can't open object
  #001: ../../../src/H5O.c line 1358 in H5O_open_name(): object not found
    major: Symbol table
    minor: Object not found
  #002: ../../../src/H5Gloc.c line 430 in H5G_loc_find(): can't find object
    major: Symbol table
    minor: Object not found
  #003: ../../../src/H5Gtraverse.c line 861 in H5G_traverse(): internal path traversal failed
    major: Symbol table
    minor: Object not found
  #004: ../../../src/H5Gtraverse.c line 755 in H5G_traverse_real(): component not found
    major: Symbol table
    minor: Object not found
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_GetAttributeNames                                           OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_GetDataDimensions                                           OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_GetFileName                                                 OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_GetLinkNames                                                OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 439 in H5Fis_hdf5(): unable open file
    major: File accessibilty
    minor: Not an HDF5 file
  #001: ../../../src/H5Fint.c line 554 in H5F_is_hdf5(): unable to open file
    major: Low-level I/O
    minor: Unable to initialize object
  #002: ../../../src/H5FD.c line 993 in H5FD_open(): open failed
    major: Virtual File Layer
    minor: Unable to initialize object
  #003: ../../../src/H5FDsec2.c line 339 in H5FD_sec2_open(): unable to open file: name = 'test.h5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0
    major: File accessibilty
    minor: Unable to open file
TestHDF5File_HDF5File1                                                   OK.

TestHDF5File_HDF5File2                                                   OK.

TestHDF5File_HDF5File3                                                   OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5A.c line 1692 in H5Aexists_by_name(): unable to determine if attribute exists
    major: Attribute
    minor: Can't get value
  #001: ../../../src/H5Aint.c line 1200 in H5A_exists_by_name(): object not found
    major: Attribute
    minor: Object not found
  #002: ../../../src/H5Gloc.c line 430 in H5G_loc_find(): can't find object
    major: Symbol table
    minor: Object not found
  #003: ../../../src/H5Gtraverse.c line 861 in H5G_traverse(): internal path traversal failed
    major: Symbol table
    minor: Object not found
  #004: ../../../src/H5Gtraverse.c line 755 in H5G_traverse_real(): component not found
    major: Symbol table
    minor: Object not found
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_HasAttribute                                                OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5D.c line 358 in H5Dopen2(): not found
    major: Dataset
    minor: Object not found
  #001: ../../../src/H5Gloc.c line 430 in H5G_loc_find(): can't find object
    major: Symbol table
    minor: Object not found
  #002: ../../../src/H5Gtraverse.c line 861 in H5G_traverse(): internal path traversal failed
    major: Symbol table
    minor: Object not found
  #003: ../../../src/H5Gtraverse.c line 755 in H5G_traverse_real(): component not found
    major: Symbol table
    minor: Object not found
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_HasFloatDataType                                            OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5D.c line 358 in H5Dopen2(): not found
    major: Dataset
    minor: Object not found
  #001: ../../../src/H5Gloc.c line 430 in H5G_loc_find(): can't find object
    major: Symbol table
    minor: Object not found
  #002: ../../../src/H5Gtraverse.c line 861 in H5G_traverse(): internal path traversal failed
    major: Symbol table
    minor: Object not found
  #003: ../../../src/H5Gtraverse.c line 755 in H5G_traverse_real(): component not found
    major: Symbol table
    minor: Object not found
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_HasIntDataType                                              OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_HasPath                                                     OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_IsDataSet                                                   OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_IsGroup                                                     OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5A.c line 347 in H5Acreate_by_name(): object not found
    major: Attribute
    minor: Object not found
  #001: ../../../src/H5Gloc.c line 430 in H5G_loc_find(): can't find object
    major: Symbol table
    minor: Object not found
  #002: ../../../src/H5Gtraverse.c line 861 in H5G_traverse(): internal path traversal failed
    major: Symbol table
    minor: Object not found
  #003: ../../../src/H5Gtraverse.c line 641 in H5G_traverse_real(): traversal operator failed
    major: Symbol table
    minor: Callback failed
  #004: ../../../src/H5Gloc.c line 385 in H5G_loc_find_cb(): object 'foo' doesn't exist
    major: Symbol table
    minor: Object not found
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5A.c line 352 in H5Acreate_by_name(): unable to create attribute
    major: Attribute
    minor: Unable to initialize object
  #001: ../../../src/H5Aint.c line 167 in H5A_create(): attribute already exists
    major: Attribute
    minor: Object already exists
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5A.c line 476 in H5Aopen_by_name(): can't open attribute
    major: Attribute
    minor: Can't open object
  #001: ../../../src/H5Aint.c line 451 in H5A_open_by_name(): unable to load attribute info from object header
    major: Attribute
    minor: Unable to initialize object
  #002: ../../../src/H5Oattribute.c line 537 in H5O_attr_open_by_name(): can't locate attribute: 'PROPERTYY'
    major: Attribute
    minor: Object not found
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadAttribute1                                              OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadAttribute10                                             OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadAttribute11                                             OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadAttribute2                                              OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadAttribute3                                              OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5A.c line 347 in H5Acreate_by_name(): object not found
    major: Attribute
    minor: Object not found
  #001: ../../../src/H5Gloc.c line 430 in H5G_loc_find(): can't find object
    major: Symbol table
    minor: Object not found
  #002: ../../../src/H5Gtraverse.c line 861 in H5G_traverse(): internal path traversal failed
    major: Symbol table
    minor: Object not found
  #003: ../../../src/H5Gtraverse.c line 641 in H5G_traverse_real(): traversal operator failed
    major: Symbol table
    minor: Callback failed
  #004: ../../../src/H5Gloc.c line 385 in H5G_loc_find_cb(): object 'foo' doesn't exist
    major: Symbol table
    minor: Object not found
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5A.c line 352 in H5Acreate_by_name(): unable to create attribute
    major: Attribute
    minor: Unable to initialize object
  #001: ../../../src/H5Aint.c line 167 in H5A_create(): attribute already exists
    major: Attribute
    minor: Object already exists
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5A.c line 476 in H5Aopen_by_name(): can't open attribute
    major: Attribute
    minor: Can't open object
  #001: ../../../src/H5Aint.c line 451 in H5A_open_by_name(): unable to load attribute info from object header
    major: Attribute
    minor: Unable to initialize object
  #002: ../../../src/H5Oattribute.c line 537 in H5O_attr_open_by_name(): can't locate attribute: 'PROPERTYY'
    major: Attribute
    minor: Object not found
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadAttribute4                                              OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadAttribute5                                              OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadAttribute6                                              OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5A.c line 347 in H5Acreate_by_name(): object not found
    major: Attribute
    minor: Object not found
  #001: ../../../src/H5Gloc.c line 430 in H5G_loc_find(): can't find object
    major: Symbol table
    minor: Object not found
  #002: ../../../src/H5Gtraverse.c line 861 in H5G_traverse(): internal path traversal failed
    major: Symbol table
    minor: Object not found
  #003: ../../../src/H5Gtraverse.c line 641 in H5G_traverse_real(): traversal operator failed
    major: Symbol table
    minor: Callback failed
  #004: ../../../src/H5Gloc.c line 385 in H5G_loc_find_cb(): object 'foo' doesn't exist
    major: Symbol table
    minor: Object not found
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5A.c line 352 in H5Acreate_by_name(): unable to create attribute
    major: Attribute
    minor: Unable to initialize object
  #001: ../../../src/H5Aint.c line 167 in H5A_create(): attribute already exists
    major: Attribute
    minor: Object already exists
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5A.c line 476 in H5Aopen_by_name(): can't open attribute
    major: Attribute
    minor: Can't open object
  #001: ../../../src/H5Aint.c line 451 in H5A_open_by_name(): unable to load attribute info from object header
    major: Attribute
    minor: Unable to initialize object
  #002: ../../../src/H5Oattribute.c line 537 in H5O_attr_open_by_name(): can't locate attribute: 'PROPERTYY'
    major: Attribute
    minor: Object not found
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadAttribute7                                              OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadAttribute8                                              OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadAttribute9                                              OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadDataSet1                                                OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadDataSet2                                                OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadDataSet3                                                OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadDataSet4                                                OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadDataSet5                                                OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadDataSet6                                                OK.

HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: ../../../src/H5F.c line 781 in H5Fclose(): invalid file identifier
    major: Invalid arguments to routine
    minor: Inappropriate type
TestHDF5File_ReadDataSet7                                                OK.

TestHDF5ModelPartIO_Properties                                           OK.

TestHDF5ModelPartIO_ReadConditions1                                      OK.

TestHDF5ModelPartIO_ReadConditions2                                      OK.

TestHDF5ModelPartIO_ReadElements1                                        OK.

TestHDF5ModelPartIO_ReadElements2                                        OK.

TestHDF5ModelPartIO_ReadNodes                                            OK.

TestHDF5PointsData_CreateNodes                                           OK.

TestHDF5PointsData_ReadData                                              OK.

TestHDF5PointsData_ReadNodalResults                                      OK.

TestHDF5PointsData_SetData                                               OK.

TestHDF5PointsData_WriteData                                             OK.

TestHDF5Utils_IsPath                                                     OK.

TestHDF5Utils_Split                                                      OK.

Ran 52 of 274 test cases in 0.0624833s. OK.
Finished running cpp unit tests!
Running python tests ...
..
----------------------------------------------------------------------
Ran 2 tests in 0.033s

OK

~~~

RiccardoRossi commented on applications/HDF5Application/custom_io/hdf5_connectivities_data.cpp:
```diff
@@ -0,0 +1,191 @@
+{
+    KRATOS_TRY;
+    Clear();
+    rFile.ReadAttribute(rPath, "Name", mName);
```
one question: what happens if you have to read multiple elements types?

msandre replied:
Connectivities data only represents a single element type. If the model part has more than one element or condition type, then a new connectivities data object is created for each type. Each instance is then responsible for writing its data to the file with the label given by the element or condition type. An example (with only one element type) is shown in the figure. 
![connectivities](https://user-images.githubusercontent.com/22290001/36165969-7d3329c6-10f1-11e8-9a8b-9b25afb225b2.png)

RiccardoRossi replied:
what about submodelparts?

msandre replied:
Sub model parts are on the todo list, but not there yet.

RiccardoRossi commented on applications/HDF5Application/custom_io/hdf5_model_part_io_base.h:
```diff
@@ -72,16 +72,12 @@ class ModelPartIOBase : public IO
 
```
I am a bit lost on why you define a HDF5 namespace with the modelpart_io base class in it. It will hide the kratos one no?

msandre replied:
Model part io is specific to hdf5. I put hdf5 specific classes in the hdf5 namespace to avoid having HDF5 at the start of every class name and to distinguish between classes, which are specific to hdf5 (know about hdf5 file for example), and others, which only know about / operate on regular kratos objects. The model part io base class was the solution used to avoid code duplication in the modelpart io and partitioned modelpart io. It should not act as the interface to the model part io. For that there is the IO class which is not in the hdf5 namespace. I'm not sure if this answers your question though.

RiccardoRossi replied:
i guess it does. It is just that i am not used to this way of doing (not a problem though)

RiccardoRossi reviewed (commented):
guys i don't know this enough to comment, so please be didactic with me: 
1 - are multiple element types or subgroups permitted in the input?
2 - is output on gauss points permitted?
3 - is there any way we could abstract the interface so that one could easily switch from GidIO to HDF5?

Commit:
added readme.

Commit:
switched to isinstance from codacy suggestion.

Commit:
codacy suggestions for xdmf.

Commit:
fix omp loops for win and add const  to function.

Commit:
using recommended error macro.

Commit:
remove obsolete tests.

Commit:
removing duplicate code in factories.

Commit:
cleaning up omp loops.

Commit:
cleaning up utility and documenting.

msandre commented:
@RiccardoRossi here is my response to the general comments.

> 1 - are multiple element types or subgroups permitted in the input?

Currently multiple element types are supported. Sub model part components are not yet supported.

> 2 - is output on gauss points permitted?

Not yet.

> 3 - is there any way we could abstract the interface so that one could easily switch from GidIO to HDF5?

I think this should be on the roadmap. I made the model part io use the IO interface in the core to make it  easy to switch, but there is no mechanism yet that I know of to select IO at runtime. The midterm solution is to do HDF5 IO through the processes. For nodal results and gp results, I think we need to develop an interface similar to IO.

Commit:
Renamed connectivities data getter methods.

Commit:
Simplified model part io base internals.

Commit:
Simplified partition utility.

Commit:
added const.

Commit:
Simplified model part io params.

Commit:
cleaned up factor elements and conditions.

jcotela commented:
Hi Mike, I'm trying to play a bit with this branch in order to do a proper review, but I just found I'm unable to run the `test_xdmf_output.py` file in `tests`. It complains about a missing `xdmf_utils` module, but I could not find it, neither in the `python_scripts` folder nor on the repositories (for python and manjaro linux). Is this file missing?

msandre commented:
Hey Jordi, yes I forgot to update the file `test_xdmf_output.py` because I didn't include it in test_HDF5Application.py. I removed it in the last set of commits (yesterday evening) because the tests are obsolete now. You can ignore it.

Commit:
Added comments and small cleaning.

Commit:
fix compile error.

Commit:
Updating documentation.

Commit:
Replaced boost::timer with BuiltinTimer based on chrono.

Commit:
small.

msandre commented:
@philbucher the boost::timer has been replaced by a kratos/utilities/builtin_timer.h. Otherwise I was just tidying up some things based on the feedback. I could not find a way to suppress the hdf5 error messages. It is a good idea about the constitutive law, but I think we should do it in a future pull request. Otherwise I think I have addressed all your points.

philbucher reviewed (dismissed):
@msandre looks good to me.
Especially thx for the timer, I think this will be handy also for others. Does it work in omp?
Also thx for the readme. Maybe in the future you can also include some info abt how to use (i.e. which processes should be used etc...)
I agree that constitutive law and other features should be in a separate PR

Commit:
fixed SetData for case of an empty container.

Commit:
fixed parallel WriteDataSet for case of empty local data set.

Commit:
fixed bugs in element and condition factoring.

Commit:
Added a condition type that only exists  on one process to mpi io test.

msandre commented:
Thanks @jcotela for the example. There was a major bug in the factoring routines which apparently went unnoticed by the existing mpi test. After fixing the bug it uncovered some additional logic errors in this pr so I had to go back and update the ConnectivitiesData and fix WriteDataSet for the parallel file. The bugs were exposed by having a condition type (it could also be an element) that exists on one process but not on another. I updated the builtin mpi test to include this case. The example you sent me should work now.

msandre commented:
@philbucher a very reliable source (the internet) told me it is thread-safe (http://www.modernescpp.com/index.php/sleep-and-wait). I tested it a bit in some omp loops and it seemed to work.

RiccardoRossi commented:
@philbucher @msandre , take care on this. The interoperability between OpenMP and c++11 threading is NOT guaranteed. For example it might work with gcc and break with another compiler. This is actually the reason for which i was pushing #943

msandre commented:
@RiccardoRossi I searched some more. I'm not sure it helps but here is what I could find. From _The C++ Programming Language 4th edition_, iso.20.11.7 on steady_clock:

> A clock where time moves steadily forward; that is, time does not go backward and the time between clock ticks is constant.

From iso.20.11.7, a steady clock must satisfy c.now() <= c.now() for consecutive calls and now() is noexcept.

I'm not calling in a thread, but do you think we should add a macro to use omp's method instead if openmp is available since it is in the core?

RiccardoRossi commented:
well this is easily measured by a small test... the killer here is the
difference between user time ans real time

the real time is the time you perceive, the other one is the sum of the
times spent in the various threads. At least some of the timers measure the
second, i learnt this the hard way. Now to be clear i am not sure if this
is the case for the std::chrono timer

cheers
Riccardo

On Thu, Feb 15, 2018 at 9:25 PM, Michael Andre <notifications@github.com>
wrote:

> @RiccardoRossi <https://github.com/riccardorossi> I searched some more.
> I'm not sure it helps but here is what I could find. From *The C++
> Programming Language 4th edition*, iso.20.11.7 on steady_clock:
>
> A clock where time moves steadily forward; that is, time does not go
> backward and the time between clock ticks is constant.
>
> From iso.20.11.7, a steady clock must satisfy c.now() <= c.now() for
> consecutive calls and now() is noexcept.
>
> I'm not calling in a thread, but do you think we should add a macro to use
> omp's method instead if openmp is available since it is in the core?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/1478#issuecomment-366050741>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHr7EbVwjGr0wuuVhajP4b5FS439nDTDks5tVJKygaJpZM4SAS8Q>
> .
>



-- 


*Riccardo Rossi*

PhD, Civil Engineer


member of the Kratos Team: www.cimne.com/kratos

Tenure Track Lecturer at Universitat Politècnica de Catalunya,
BarcelonaTech (UPC)

Full Research Professor at International Center for Numerical Methods in
Engineering (CIMNE)


C/ Gran Capità, s/n, Campus Nord UPC, Ed. B0, Despatx 102

(please deliver post to the porters of building C1)

08034 – Barcelona – Spain – www.cimne.com  -

T.(+34) 93 401 56 96 skype: *rougered4*



<http://www.cimne.com/>

<https://www.facebook.com/cimne> <http://blog.cimne.com/>
<http://vimeo.com/cimne> <http://www.youtube.com/user/CIMNEvideos>
<http://www.linkedin.com/company/cimne> <https://twitter.com/cimne>

Les dades personals contingudes en aquest missatge són tractades amb la
finalitat de mantenir el contacte professional entre CIMNE i voste. Podra
exercir els drets d'accés, rectificació, cancel·lació i oposició,
dirigint-se a cimne@cimne.upc.edu. La utilització de la seva adreça de
correu electronic per part de CIMNE queda subjecte a les disposicions de la
Llei 34/2002, de Serveis de la Societat de la Informació i el Comerç
Electronic.

 Imprimiu aquest missatge, només si és estrictament necessari.
<http://www.cimne.com/>

msandre commented:
Ok, I agree it's important to check different compiler vendors in addition to the standard. I tried the following test on gnu compiler, clang and visual studio:

```
int main()
{
  BuiltinTimer timer;
  std::array<bool, 7> flag{};
  double x{};
  std::cout << std::setprecision(12);

#pragma omp parallel default(shared) reduction(+ : x)
  {
    const int thread_id = omp_get_thread_num();
    std::cout << "Hello from thread " << thread_id << std::endl << std::flush;
    #pragma omp barrier
    if (thread_id == 0)
      std::cout << "omp_get_wtime()\t\t steady_clock::now()" << std::endl;
    const double t0 = omp_get_wtime();
    x = { 1.0 };
    while (omp_get_wtime() - t0 < 60.0)
      {
       	const unsigned i = omp_get_wtime() - t0;
	const double elapsed = timer.ElapsedSeconds();
	if (!flag.at(i / 10) && thread_id == 0)
          {
            std::cout << omp_get_wtime() - t0 << "\t\t" << elapsed << std::endl;
            flag.at(i / 10) = true;
          }
	else
          {
            x = std::cos(x); // Do work.
          }
      }
  }
  std::cout << "Finished in " << timer.ElapsedSeconds() << " seconds!" << std::endl;
  std::cout << x << std::endl;
  return 0;
}
```

There is a slight shift in time due to the steady_clock() starting before the omp threads are spawned.

g++ 5.4:
`g++ -std=c++11 -pedantic-errors -Wall -Wextra -fopenmp -o main main.cpp && ./main`
```
Hello from thread Hello from thread 1
Hello from thread 0
Hello from thread 32

omp_get_wtime()		 steady_clock::now()
4.4098123908e-07		9.186e-05
10.000000209		10.000091785
20.000000155		20.000091725
30.000017037		30.000108575
40.000000107		40.000091688
50.005308478		50.005400047
Finished in 60.000150284 seconds!
2.95634053286
```

clang 5.0:
`clang++-5.0 -std=c++11 -pedantic -Wall -fopenmp -o main main.cpp && ./main`
```
Hello from thread Hello from thread 01
Hello from thread 3
Hello from thread 2

omp_get_wtime()		 steady_clock::now()
0		0.000383482
10.0000040531		10.000387786
20.0068640709		20.007247466
30		30.000383016
40		40.000382926
50.0196368694		50.02002054
60		60.000382924
Finished in 60.000511807 seconds!
2.95634053286
```

Visual Studio 2017:

```
Hello from thread Hello from thread 3
0
Hello from thread 2
Hello from thread 1
omp_get_wtime()          steady_clock::now()
0               0.322132462
10.0161767485           10.338309211
20              20.322132462
30.0160853944           30.338217857
40              40.322132462
50              50.322132462
Finished in 60.322140577 seconds!
2.95634053286
```

Everything I've read indicates the chrono clocks go by physical time. Also boost says it about c++11 steady clocks:

> Steady clocks are defined by the C++11 standard as clocks for which values never decrease as physical time advances and for which values advance at a steady rate relative to real time. That is, the clock may not be adjusted. Clocks that are steady never run backwards, even when the operating system's clock is reset backwards such as during a daylight saving time transition.

I used the pedantic flag to try to warn about the use of language extensions whenever possible. 

Unless there are more possible problems which I overlooked, I would suggest we leave BuiltinTimer like it is.

jcotela reviewed (approved):
I'm approving since I finally managed to get my tests to write valid HDF5/XDMF output using this branch.

I do not have major objections with the code. I'm OK with the global organization and you have already discussed the details with @RiccardoRossi and @philbucher.

I agree with @RiccardoRossi that we need to write a wrapper with the same interface as the gid output, but this can be addressed in a future PR.
